src/rtt.cpp

```diff
@@ -0,0 +1,590 @@
+        highLevelFunctions.dll_close();
+
+        releaseHighLevel();
+
+        uint32_t clockSpeed = probeInfo.clockspeed_khz;
+        device_family_t family = deviceInfo.device_family;
+        std::string jlinkarmlocation = libraryInfo.file_path;
+
+        RETURN_ERROR_ON_FAIL((nrfjprogdll_err_t)loadnRFjprogFunctions(&dll_function), RTTCouldNotLoadnRFjprogLibrary);
+
+        libraryLoaded = true;
+
+        RETURN_ERROR_ON_FAIL(dll_function.open_dll(jlinkarmlocation.c_str(), &RTT::logCallback, family), RTTCouldNotOpennRFjprogLibrary);
+
+        RETURN_ERROR_ON_FAIL(dll_function.connect_to_emu_with_snr(baton->serialNumber, clockSpeed), RTTCouldNotConnectToDevice);
+        RETURN_ERROR_ON_FAIL(dll_function.connect_to_device(), RTTCouldNotConnectToDevice);
+
+        if (baton->hasControlBlockLocation) {
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_set_control_block_address(baton->controlBlockLocation), RTTCouldNotFindControlBlock);
+        }
+
+        rttStartTime = std::chrono::high_resolution_clock::now();
+        RETURN_ERROR_ON_FAIL(dll_function.rtt_start(), RTTCouldNotStartRTT);
+
+
+        while (true) {
+            bool controlBlockFound = false;
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_is_control_block_found(&controlBlockFound), RTTCouldNotCallFunction);
+            std::chrono::high_resolution_clock::time_point attemptedStartupTime = std::chrono::high_resolution_clock::now();
+
+            if (controlBlockFound) {
+                break;
+            }
+
+            if (std::chrono::duration_cast<std::chrono::seconds>(attemptedStartupTime - rttStartTime).count() > 10) {
+                cleanup();
+                return RTTCouldNotFindControlBlock;
+            }
+        }
+
+        uint32_t downChannelNumber;
+        uint32_t upChannelNumber;
+
+        RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_count(&downChannelNumber, &upChannelNumber), RTTCouldNotGetChannelInformation);
+
+        for(uint32_t i = 0; i < downChannelNumber; ++i)
+        {
+            char channelName[32];
+            uint32_t channelSize;
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_info(i, DOWN_DIRECTION, channelName, 32, &channelSize), RTTCouldNotGetChannelInformation);
+
+            std::string name(channelName);
+            baton->downChannelInfo.push_back(new ChannelInfo(i, name, channelSize));
+        }
+
+        for(uint32_t i = 0; i < upChannelNumber; ++i)
+        {
+            char channelName[32];
+            uint32_t channelSize;
+            RETURN_ERROR_ON_FAIL(dll_function.rtt_read_channel_info(i, UP_DIRECTION, channelName, 32, &channelSize), RTTCouldNotGetChannelInformation);
+
+            std::string name(channelName);
+            baton->upChannelInfo.push_back(new ChannelInfo(i, name, channelSize));
+        }
+
+        return RTTSuccess;
+    };
+
+    rtt_return_function_t r = [&] (RTTBaton *b) -> returnType {
```

Comment:
`returnType` might be a magic that the compiler solves here, but does not allow a readable code. Also again `vector` is a name that I would avoid using.